MantidPlot/src/ProjectRecovery.cpp

```diff
@@ -630,6 +632,8 @@ void ProjectRecovery::saveWsHistories(const Poco::Path &historyDestFolder) {
   std::vector<boost::shared_ptr<Mantid::API::Workspace>> wsHandles =
       ads.getObjects(Mantid::Kernel::DataServiceHidden::Include);
 
+  removeEmptyGroupsFromADS(wsHandles);
```

Comment:
I think this should be done after the check that the handles are empty. Seems more logical. Move below following statement.

Reply:
Nevertheless, this can be done as part of the existing loop lines 651. The dyanmic_cast and count-check should IMHO be performed there. That would eliminate the need for a secondary loop and the need to copy the vector of shared_ptr handles around.

Reply:
Check is performed before the empty check on purpose, so that it can remove groups before the check, because we want it to act as if it were empty should only empty groups remain.

Reply:
Also I don't copy them around I pass them by reference so that I can edit it on the fly, instead of making a copy.

Reply:
Maybe I'm missing something, but why wouldn't something along the lines of this be a better solution? As part of [this](https://github.com/mantidproject/mantid/blob/master/MantidPlot/src/ProjectRecovery.cpp#L646-L661) block
```cpp
for (auto i = 0u; i < wsHandles.size(); ++i) {
  if(auto groupWS = boost::dynamic_pointer_cast<Mantid::API::WorkspaceGroup>(wsHandles[i])) {
    if(groupWS->isEmpty())
      continue;
}
...
```

Reply:
Ah, this is something different to how I interpreted your earlier responses, that is actually a good idea! However previously I believed you were asking about the check on line 636 where a check for wsHandles.empty() is performed.